### PR TITLE
Weak referenced Identity cache in the CurrentUser class

### PR DESCRIPTION
**Pull request by roxblnfk**
| Q             | A
| ------------- | ---
| Fixed issues  | #2 

An example of how you can store the cache using WeakReference

### CODE REVIEW

File: src/CurrentUser.php [hunk 1]
```diff
@@ -76,7 +77,9 @@ public function setAbsoluteAuthTimeout(int $timeout = null): self
      */
     public function getIdentity(): IdentityInterface
     {
-        $identity = $this->temporaryIdentity ?? $this->identity;
+        /** @var IdentityInterface|null $identity */
+        $identity = ($this->temporaryIdentity !== null ? $this->temporaryIdentity->get() : null)
+            ?? ($this->identity !== null ? $this->identity->get() : null);
```
Comment (xepozz) [inline src/CurrentUser.php:82]:
```suggestion
        $identity = $this->temporaryIdentity->get() ?? $this->identity->get();
```

Comment (roxblnfk) [reply src/CurrentUser.php:82]: Error: Call to a member function get() on null

**Comment by samdark** [comment]
https://github.com/yiisoft/user/pull/28

**Comment by samdark** [comment]
It won't function well if the identity is clean and isn't tied to any ORM i.e. reposiory is using just SQL and `new Identity($data)`.

**Comment by samdark** [comment]
See https://3v4l.org/Cjirn

**Comment by roxblnfk** [comment]
> See https://3v4l.org/Cjirn
 
did you forget that after this [PR](https://github.com/yiisoft/user/pull/28) i was going to change the way the `statedIdetity` is stored inside the User service

**Comment by samdark** [comment]
@roxblnfk but this is the same for primary identity as well. The difference is that it will be read from the storage each time it is requested and that isn't good. For example:

```php
<?php foreach ($posts as $post): ?>
   <?php if ($currentUser->can('edit_post')): ?>
       <a href="#">Edit post</a>
   <?php endif ?>
   ...
<?php endforeach ?>
```

If you're rendering such list of a 100 posts, you'll get 100 readings from identity storage.

**Comment by roxblnfk** [comment]
> If you're rendering such list of a 100 posts, you'll get 100 readings from identity storage.

you need to try very hard to fulfill all the conditions so that the current user does not cached in the service by this moment.
however, if you like to store the state and not even separate the implementation of the cache storage, then i, perhaps, will leave these useless attempts

**Comment by samdark** [comment]
> you need to try very hard to fulfill all the conditions so that the current user does not cached in the service by this moment.

No, that's easy:

```php
class UserIdentity implements IdentityInterface
{
    private string $id;
    private string $login;
    private string $password;

    public function __construct(string $id, array $attributes)
    {
        $this->id = $id;
        $this->login = $attributes['login'] ?? null;
        $this->password = $attributes['password'] ?? null;
    }

    public function getId(): ?string
    {
        return $this->id;
    }
}

class UserRepository implements IdentityRepositoryInterface
{
    private array $users = [
        1 => [
            'login' => 'Doris',
            'password' => 'qwerty',
        ],
        2 => [
            'login' => 'Boris',
            'password' => 'ytrewq',
        ],
    ];

    public function findIdentity(string $id): ?IdentityInterface
    {
        // could be getting identity from DB via SQL + PDO
        if (!array_key_exists($id, $this->users)) {
            return null;
        }

        return new UserIdentity($id, $this->users[$id]);
    }
}

```

**Pull request closed without merging**